L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h

```diff
@@ -0,0 +1,152 @@
+#ifndef L1Trigger_L1TTrackMatch_L1TkEtMissEmuAlgo_HH
+#define L1Trigger_L1TTrackMatch_L1TkEtMissEmuAlgo_HH
+
+#include "DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h"
+#include "DataFormats/L1TrackTrigger/interface/TTTypes.h"
+#include <numeric>
+
+#include <cmath>
+#include <iostream>
+#include <cstdint>
+#include <filesystem>
+#include <fstream>
+#include <ap_int.h>
+
+using namespace std;
+// Namespace that defines constants and types used by the EtMiss Emulation
+// Includes functions for writing LUTs and converting to integer representations
+namespace L1TkEtMissEmuAlgo {
+
+  const unsigned int N_chi2XYbits_{4};  //Taken from TTTrack_Word class
+  const unsigned int N_chi2bendbits_{3};
+  const unsigned int N_hitpatternbits_{7};
+  const unsigned int N_ptBits_{15};
+  const unsigned int N_etaBits_{16};
+  const unsigned int N_z0Bits_{12};
+  const unsigned int N_phiBits_{12};
+
+  const unsigned int N_globphiBits_{11};  //Precision of Cos LUT (only spans 0 to pi/2) whereas global phi ( 0 to 2 pi)
+  const unsigned int N_globphiExtra_{3};  //Extra bits needed by global phi to span full range
+  const unsigned int N_EtExtra_{8};       //Extra room for Et sums
+
+  const unsigned int N_METbits_{14};     //For output Magnitude
+  const unsigned int N_METphibits_{11};  //For Output Phi
+
+  typedef ap_uint<N_chi2XYbits_> ichi2XY;
+  typedef ap_uint<N_chi2bendbits_> ichi2bend;
+  typedef ap_uint<3> iNstub;
+  typedef ap_uint<N_phiBits_> iPhi;
+  typedef ap_uint<N_globphiBits_ + N_globphiExtra_> iglobPhi;
+  typedef ap_uint<N_z0Bits_> iZ0;
+  typedef ap_uint<N_ptBits_> iPt;
+  typedef ap_int<N_ptBits_ + N_EtExtra_> iEt;
+  typedef ap_uint<N_etaBits_> iEta;
+
+  typedef ap_uint<N_METbits_> iMET;
+  typedef ap_uint<N_METphibits_>
+      iMETphi;  //Cordic means this is evaluated between 0 and 2Pi rather than -pi to pi so unsigned
+
+  /*
```

Comment:
You'll probably want to remove any code which is commented out.

Reply:
Fixed in latest commit